invites: Fix bug revoking user invites in other realms than intended.

mateuszmandera opened the pull request:
Fixes #17238.
In process_new_human user, the queries were wrong, revoking all invites
sent to the email address, even in other realms than the one where the
new account just got created.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -463,18 +463,21 @@ def process_new_human_user(user_profile: UserProfile,
-        PreregistrationUser.objects.filter(email__iexact=user_profile.delivery_email)\
+    elif prereg_user is None:
+        assert not realm_creation
+        PreregistrationUser.objects.filter(email__iexact=user_profile.delivery_email, realm=user_profile.realm)\
```
I don't understand -- with `realm_creation`, don't we want `realm=None`?

mateuszmandera replied:
realm_creation should never happen without a prereg_user being provided - after all we're coming from a confirmation link for realm creation. And in that case we want to activate this specific prereguser and leave all other ones alone - the email address might have requested two realm creation links because they want to create two organization. I added a commit with a test for such a scenario.

Also the `realm_creation should never happen without a prereg_user being provided ` assumption should be more clear in the extracted revoke_... function.

timabbott replied:
I still don't understand -- what I'd expect is that `used_pregistration_user` would have `realm=None` attached to it, and thus I expect this query to do nothing:
```
def create_preregistration_user(email: str, request: HttpRequest, realm_creation: bool=False,                                        
                                password_required: bool=True, full_name: Optional[str]=None,                                         
                                full_name_validated: bool=False) -> HttpResponse:                                                    
    realm = None                                                                                                                     
    if not realm_creation:                                                                                                           
        try:                                                                                                                         
            realm = get_realm(get_subdomain(request))                                                                                
        except Realm.DoesNotExist:                                                                                                   
            pass                                                                                                                     
    return PreregistrationUser.objects.create(                                                                                       
        email=email,                                                                                                                 
        realm_creation=realm_creation,                                                                                               
        password_required=password_required,                                                                                         
        realm=realm,                                                                                                                 
        full_name=full_name,                                                                                                         
        full_name_validated=full_name_validated,                                                                                     
    )                                                                                                                                
```

Isn't the right logic here just to mutate `used_pregistration_user` in the `realm_creation` code path?

timabbott replied:
If you print them out, `prereg_user.realm` is `None` while `created_user_profile.realm` is the new realm in the `realm_creation` code path, and I'm pretty sure we want the `None`.

mateuszmandera replied:
> Isn't the right logic here just to mutate `used_pregistration_user` in the `realm_creation` code path?

That's right. This prereguser was already being mutated at the beginnng of process_new_human_user. I moved that logic into the function now because that seems less confusing. With the two new refactoring commits I think things should be more readable and easier to identify as correct behavior.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -463,18 +463,21 @@ def process_new_human_user(user_profile: UserProfile,
             notify_invites_changed(user_profile)
-    else:
-        PreregistrationUser.objects.filter(email__iexact=user_profile.delivery_email)\
+    elif prereg_user is None:
```
Should this just be nesting inside a `if prereg_user is not None` block that has an `if realm_creation` feature in it?

May be cleaner to extract to a function `revoke_preregistration_users(delivery_email, realm: Optional[Realm])`?  And the semantics would be `realm=None` iff `realm_creation`?

mateuszmandera replied:
As explained above in https://github.com/zulip/zulip/pull/17275#discussion_r575282584, `if realm_creation` is not an allowable scenario without a prereg_user. Extracting the logic for revoking PreregistrationUsers seems like a great idea to clean things up, so pushed a commit.

timabbott replied:
Cool.  I think we can do a simpler if/else sequence here, more like this:

```
def revoke_preregistration_users(                                                                    
    created_user_profile: UserProfile,                                                               
    used_preregistration_user: Optional[PreregistrationUser],                                        
    realm_creation: bool                                                                             
) -> None:                                                                                           
    if used_preregistration_user is None:              
        assert not realm_creation, "realm_creation should only happen with a PreregistrationUser"
        return
                                                                                                     
    # Mark any other PreregistrationUsers in the realm that are STATUS_ACTIVE as                     
    # inactive so we can keep track of the PreregistrationUser we                                    
    # actually used for analytics.                                                                   
    if realm_creation:                                                                               
        # In the special case of realm creation, we just need                                        
        PreregistrationUser.objects \                                                                
            .filter(email__iexact=created_user_profile.delivery_email, realm=prereg_user.realm) \                                                                                               
            .update(status=confirmation_settings.STATUS_REVOKED)                                     
    else:                                                                                            
        PreregistrationUser.objects \                                                                
            .filter(email__iexact=created_user_profile.delivery_email, realm=prereg_user.realm) \                                                                                               
            .exclude(id=used_preregistration_user.id) \                                              
            .update(status=confirmation_settings.STATUS_REVOKED)                                     
```

mateuszmandera replied:
Yup, pushed a commit

showell commented:
Does it make sense to squash the test commit here up into the commit that fixes things?

mateuszmandera commented:
> Does it make sense to squash the test commit here up into the commit that fixes things?

Yeah that seems right - I squashed it and also  moved the queue_processor code cleanup to be the first commit.

timabbott commented:
Merged, thanks @mateuszmandera!
